### PR TITLE
Publish/subscribe

### PR DESCRIPTION
**Pull request by Meggielqk**

### CODE REVIEW

File: en_US/messaging/mqtt-auto-subscription.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+```bash
+topic: client/demo_client1/username/admin/host/192.168.1.234/port/55678
+qos: 1
+rh: 0
+rap: 0
+nl: 0
+```
+
+## Try Auto Subscription using MQTT X Client
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-auto-subscription.md]:
```suggestion
## Try Auto Subscription Using MQTT X Client
```

File: en_US/messaging/mqtt-shared-subscription.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+:::tip Prerequisites
+
+- Knowledge about MQTT [Shared Subscription](./mqtt-concepts.md)
+- Basic publishing and subscribing operations using [MQTT X](./publish-and-subscribe.md) 
+
+:::
+
+## Try Shared Subscription with MQTT X Client
+
+The following procedure demonstrates how to form groups for multiple subscribers and make them subscribe to the same topic. In this demonstration, you can create three subscribers `Subscriber1`, `Subscriber2` and `Subscriber3,` and divide them into groups `a` and `b`. Both groups subscribe to the topic `t/1`.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-shared-subscription.md]:
```suggestion
The following procedure demonstrates how to group multiple subscribers and make them subscribe to the same topic. In this demonstration, you can create three subscribers `Subscriber1`, `Subscriber2` and `Subscriber3,` and divide them into groups `a` and `b`. Both groups subscribe to the topic `t/1`. 
```

Comment (LenaLenaPan) [reply en_US/messaging/mqtt-shared-subscription.md]: 这里其实还建了一个 demo connection，要不要说下这个 demo connection 的目的？

File: en_US/messaging/publish-and-subscribe.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
+# Test with MQTT Clients 
+
+It is more safe and efficient for the developers to test the EMQX messaging services using client tools before connecting the real-time devices to the EMQX broker and developing the IoT applications. If some errors are detected during the testing, the developers can troubleshoot them in the test environment rather than in production.
```
Comment (LenaLenaPan) [inline en_US/messaging/publish-and-subscribe.md]:
```suggestion
It is more safe and efficient for the developers to test the EMQX messaging services using client tools before connecting the real-time devices to the EMQX and developing the IoT applications.
```

**Review by LenaLenaPan — COMMENTED** [review]
add some comments

File: en_US/messaging/mqtt-delayed-publish.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# Delayed Publish
+
+Delayed publish is an extended MQTT feature supported by EMQX. When a client publishes a message to EMQX broker with the topic prefix `$delayed/{DelayInteval}`, it triggers the delayed publish feature. The messages will be published after a period of time predefined by the user.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-delayed-publish.md]:
```suggestion
Delayed publish is an extended MQTT feature supported by EMQX. When a client publishes a message to EMQX with the topic prefix `$delayed/{DelayInteval}`, it triggers the delayed publish feature. The messages will be published after a period of time predefined by the user. 
```

File: en_US/messaging/mqtt-retained-message.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
+# Retained Message
+
+EMQX implements the retained message feature of MQTT. You can flag one message published under a certain topic as `Retained` and save it as a persistent message on the EMQX broker. When any new subscribers subscribe to a topic that matches the topic of the retained message, they immediately receive that message, even if it was published before they subscribed to the topic.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-retained-message.md]:
```suggestion
EMQX implements the retained message feature of MQTT. You can flag one message published under a certain topic as `Retained` and save it as a persistent message on the EMQX. When any new subscribers subscribe to a topic that matches the topic of the retained message, they immediately receive that message, even if it was published before they subscribed to the topic.
```

File: en_US/messaging/mqtt-shared-subscription.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+     <img src="./assets/Receive-message-shared-subscription1.png" alt="Receive-message-shared-subscription1" style="zoom:35%;" />
+
+   - Send the same message again. The client `Subscriber2` and `Subscriber3` should receive the message.
+
+     <img src="./assets/Receive-message-shared-subscription2.png" alt="Receive-message-shared-subscription2" style="zoom:35%;" />
+
+:::tip
+
+When the message of the shared subscription is published, the EMQX broker forwards the message to different groups at the same time, but only one of the subscribers in the same group receives the message at a time.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-shared-subscription.md]:
```suggestion
When the message of the shared subscription is published, the EMQX forwards the message to different groups at the same time, but only one of the subscribers in the same group receives the message at a time.
```

File: en_US/messaging/mqtt-will-message.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# Will Message
+
+EMQX implements the will message feature of MQTT. If a will message is set for a client, EMQX sends the message to relevant subscribers when the client is accidentally disconnected, so that the subscribers can be informed and update the client status.
+
+You can use client tools to test this messaging service in EMQX. This section introduces how to use the [MQTT X Client](https://mqttx.app/) and [MQTT X CLI](https://mqttx.app/cli) to simulate clients and test how a will message is published and received.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-will-message.md]:
```suggestion
You can use client tools to try this messaging service in EMQX. This section introduces how to use the [MQTT X Client](https://mqttx.app/) and [MQTT X CLI](https://mqttx.app/cli) to simulate clients and see how a will message is published and received.
```

File: en_US/messaging/mqtt-retained-message.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+You can use client tools to test this messaging service in EMQX. This section introduces how to use the [MQTT X Client](https://mqttx.app/) and [MQTT X CLI](https://mqttx.app/cli) to simulate clients and try how a retained message is published and received.
```
Comment (LenaLenaPan) [inline en_US/messaging/mqtt-retained-message.md]:
```suggestion
You can use client tools to try this messaging service in EMQX. This section introduces how to use the [MQTT X Client](https://mqttx.app/) and [MQTT X CLI](https://mqttx.app/cli) to simulate clients and see how a retained message is published and received.
```

**Review by LenaLenaPan — COMMENTED** [review]
leave some comments

File: en_US/messaging/explore-mqtt.md [hunk 9]
```diff
@@ -1,26 +1,75 @@
...
   :::
 
-## Will message
+## Will Message
+
+A will message is the last stored message that is defined in advance and is sent by the server to the relevant subscriber when the network is disconnected unexpectedly. The will message feature is used in the following scenarios:
+
+- I/O error or network failure occurs on the server.
+- The client loses contact during a defined heartbeat period.
+- The client closes the network connection before sending offline packets.
```
Comment (wivwiv) [inline en_US/messaging/explore-mqtt.md]: Strictly speaking, it should be called DISCONNECT packet

```suggestion
- The client closes the network connection before sending DISCONNECT packet.
```

File: en_US/messaging/explore-mqtt.md [hunk 10]
```diff
@@ -1,26 +1,75 @@
...
+- The server closes the network connection before receiving the offline packet.
```
Comment (wivwiv) [inline en_US/messaging/explore-mqtt.md]:
```suggestion
- The server closes the network connection before receiving the DISCONNECT packet.
```

**Review by wivwiv — CHANGES_REQUESTED** [review]
Not yet review finished

**Review by wivwiv — APPROVED** [review]

**Pull request merged**